spec-tests/src/test/java/org/candlepin/spec/OwnerProductResourceSpecTest.java

```diff
@@ -0,0 +1,799 @@
+    }
+
+    @Test
+    @DisplayName("should return bad request when inserting product with null id")
+    public void shouldReturnBadRequestWhenInsertingProductWithNullId() throws Exception {
+        OwnerDTO owner = ownerApi.createOwner(Owners.random());
+        ProductDTO product = Products.random()
+            .id(null);
+
+        Request request = Request.from(client)
+            .setPath(OWNER_PRODUCTS_PATH)
+            .setMethod("POST")
+            .setPathParam("owner_key", owner.getKey())
+            .setBody(product);
+
+        assertEquals(400, request.execute().getCode());
+    }
+
+    private ProductDTO createProductWithProvidedProduct(String ownerKey) throws ApiException {
+        ProductDTO derivedProvProduct = ownerProductApi.createProductByOwner(ownerKey, Products.random());
+        ProductDTO derivedProduct = Products.random();
+        derivedProduct.setProvidedProducts(Set.of(derivedProvProduct));
+        derivedProduct = ownerProductApi.createProductByOwner(ownerKey, derivedProduct);
+
+        return derivedProduct;
+    }
+
+    private ProductDTO createProductWithProvidedAndDerivedProduct(String ownerKey) throws ApiException {
+        ProductDTO derivedProduct = createProductWithProvidedProduct(ownerKey);
+        ProductDTO provProduct = ownerProductApi.createProductByOwner(ownerKey, Products.random());
+        ProductDTO product = Products.random();
+        product.setProvidedProducts(Set.of(provProduct));
+        product.setDerivedProduct(derivedProduct);
+        product = ownerProductApi.createProductByOwner(ownerKey, product);
+
+        ownerApi.createPool(ownerKey, Pools.random(product));
+
+        return product;
+    }
+
+    private ProductDTO createHostedProductWithProvidedProduct() throws ApiException {
+        ProductDTO derivedProvProduct = hostedTestApi.createProduct(Products.random());
+        ProductDTO derivedProduct = Products.random();
+        derivedProduct.setProvidedProducts(Set.of(derivedProvProduct));
+        derivedProduct = hostedTestApi.createProduct(derivedProduct);
+
+        return derivedProduct;
+    }
+
+    private ProductDTO createHostedProductWithProvidedAndDerivedProduct() throws ApiException {
+        ProductDTO derivedProduct = createHostedProductWithProvidedProduct();
+        ProductDTO provProduct = hostedTestApi.createProduct(Products.random());
+        ProductDTO product = Products.random();
+        product.setProvidedProducts(Set.of(provProduct));
+        product.setDerivedProduct(derivedProduct);
+        product = hostedTestApi.createProduct(product);
+
+        return product;
+    }
+
+    private SubscriptionDTO createUpstreamSubscription(OwnerDTO owner, ProductDTO product)
+        throws ApiException {
+        SubscriptionDTO subscription = hostedTestApi.createSubscription(Subscriptions.random(owner, product));
+        AsyncJobStatusDTO job = ownerApi.refreshPools(owner.getKey(), false);
```

Comment:
Be *VERY* careful here. Automating/hiding the refresh op lead to very problem we had with the first iteration of the hosted test API, which is a problem that still infects the spec tests to this day, some 7 years later. 
This method looks especially close to the very deprecated `create_pools_and_subscriptions` method from the ruby hostedtest_api.rb, which has side effects which are sometimes unexpected and hides or performs actions that the test author and maintainers need to be aware of if the test is to work as intended.

You'll need the upstream logic bits for testing locked entities (which I don't see at present), but I feel like this could be better represented/abstracted with a set of nested tests and the common setup.

Reply:
I took a look and saw what you were doing in the owner-content spec suite and I added a nested test class that handles the creation of the upstream testing/locking.